### ISSUE TITLE
./cluster/kube-up.sh using vagrant error in kube-apiserver logs "remote error: bad certificate"

### ISSUE THREAD

**Issue by rawlingsj** [kubernetes/kubernetes#12854]
TL;DR vagrant script stuck at `Waiting for each minion to be registered with cloud provider`

Following the steps from https://github.com/kubernetes/kubernetes/blob/master/docs/getting-started-guides/vagrant.md#setup

I've downloaded the latest release (1.0.3) and run a master and single node using the vagrant image and virtualbox provider..

```sh
curl -L https://github.com/kubernetes/kubernetes/releases/download/v1.0.3/kubernetes.tar.gz | tar -zxv
cd kubernetes
export VAGRANT_DEFAULT_PROVIDER=virtualbox
export KUBERNETES_PROVIDER=vagrant
./cluster/kube-up.sh
```
The vagrant script remains stuck at `Waiting for each minion to be registered with cloud provider`

Kubelet and docker are running fine on both master and minion-1

Images have been pulled and are running on master node..
```sh
[root@kubernetes-master kubernetes]# docker ps
CONTAINER ID        IMAGE                                                                               COMMAND                CREATED             STATUS              PORTS               NAMES
14ee93fe4bef        gcr.io/google_containers/etcd:2.0.12                                                "/bin/sh -c '/usr/lo   19 minutes ago      Up 19 minutes                           k8s_etcd-container.fa2ab1d9_etcd-server-kubernetes-master_kube-system_6ef1b07678064a7edcdbec6f0570e917_01002528
bca706d75ba0        gcr.io/google_containers/kube-apiserver:d701e2919c33fe39b33b52fb5796ed79            "/bin/sh -c '/usr/lo   19 minutes ago      Up 19 minutes                           k8s_kube-apiserver.177ef603_kube-apiserver-kubernetes-master_kube-system_23a913d193df4488a2c2ca2e95d6c272_e6aa9ba1
810ca3b68446        gcr.io/google_containers/kube-controller-manager:1cb70d2fbd0639943228776462a3b8d6   "/bin/sh -c '/usr/lo   19 minutes ago      Up 19 minutes                           k8s_kube-controller-manager.d7e00aba_kube-controller-manager-kubernetes-master_kube-system_49cbcee20d3747842d98af77114875f8_5a22315e
435002d94066        gcr.io/google_containers/kube-scheduler:ef914b9ba028414e9e12edc42566987f            "/bin/sh -c '/usr/lo   19 minutes ago      Up 19 minutes                           k8s_kube-scheduler.bda3c268_kube-scheduler-kubernetes-master_kube-system_957055a2dbf6304dde0f132eec752c69_3dee41f3
c6c51261fd0f        gcr.io/google_containers/pause:0.8.0                                                "/pause"               19 minutes ago      Up 19 minutes                           k8s_POD.e4cc795_kube-apiserver-kubernetes-master_kube-system_23a913d193df4488a2c2ca2e95d6c272_95de218d
87c961043940        gcr.io/google_containers/pause:0.8.0                                                "/pause"               19 minutes ago      Up 19 minutes                           k8s_POD.e4cc795_etcd-server-kubernetes-master_kube-system_6ef1b07678064a7edcdbec6f0570e917_bd317a37
9c159856bc59        gcr.io/google_containers/pause:0.8.0                                                "/pause"               19 minutes ago      Up 19 minutes                           k8s_POD.e4cc795_kube-scheduler-kubernetes-master_kube-system_957055a2dbf6304dde0f132eec752c69_4741e275
60ba07473aa4        gcr.io/google_containers/pause:0.8.0                                                "/pause"               19 minutes ago      Up 19 minutes                           k8s_POD.e4cc795_kube-controller-manager-kubernetes-master_kube-system_49cbcee20d3747842d98af77114875f8_c301cefa
```
looking at the `kube-apiserver.log` there are certificate errors..

tail -f /var/log/kube-apiserver.log
```
[root@kubernetes-master kubernetes]# tail -f /var/log/kube-apiserver.log
I0818 11:48:29.653931       6 handlers.go:137] GET /api/v1/persistentvolumes: (1.450176ms) 200 [[kube-controller-manager/v1.0.3 (linux/amd64) kubernetes/61c6ac5] 127.0.0.1:44533]
I0818 11:48:30.146257       6 handlers.go:137] GET /api: (43.571µs) 200 [[kubectl/v1.0.3 (darwin/amd64) kubernetes/61c6ac5] 10.245.1.1:56098]
I0818 11:48:30.247145       6 logs.go:41] http: TLS handshake error from 10.245.1.3:47703: remote error: bad certificate
I0818 11:48:30.247680       6 logs.go:41] http: TLS handshake error from 10.245.1.3:47704: remote error: bad certificate
I0818 11:48:30.248927       6 logs.go:41] http: TLS handshake error from 10.245.1.3:47702: remote error: bad certificate
I0818 11:48:30.277062       6 handlers.go:137] GET /api: (58.954µs) 200 [[kubectl/v1.0.3 (darwin/amd64) kubernetes/61c6ac5] 10.245.1.1:56099]
I0818 11:48:30.441006       6 handlers.go:137] GET /api/v1/nodes: (1.339712ms) 200 [[kubectl/v1.0.3 (darwin/amd64) kubernetes/61c6ac5] 10.245.1.1:56100]
I0818 11:48:31.255287       6 logs.go:41] http: TLS handshake error from 10.245.1.3:47708: remote error: bad certificate
I0818 11:48:31.256085       6 logs.go:41] http: TLS handshake error from 10.245.1.3:47707: remote error: bad certificate
I0818 11:48:31.258167       6 logs.go:41] http: TLS handshake error from 10.245.1.3:47706: remote error: bad certificate
I0818 11:48:31.697763       6 handlers.go:137] GET /api/v1/namespaces/kube-system/pods/kube-ui-v1-dsdvw: (1.382056ms) 200 [[kube-scheduler/v1.0.3 (linux/amd64) kubernetes/61c6ac5] 127.0.0.1:42139]
I0818 11:48:31.711506       6 handlers.go:137] PUT /api/v1/namespaces/kube-system/events/kube-ui-v1-dsdvw.13fb8af2d2069a96: (12.070405ms) 200 [[kube-scheduler/v1.0.3 (linux/amd64) kubernetes/61c6ac5] 127.0.0.1:42139]
I0818 11:48:31.762934       6 handlers.go:137] GET /api/v1/namespaces/kube-system/pods/kube-dns-v8-1rpl3: (2.432084ms) 200 [[kube-scheduler/v1.0.3 (linux/amd64) kubernetes/61c6ac5] 127.0.0.1:42139]
I0818 11:48:31.777054       6 handlers.go:137] PUT /api/v1/namespaces/kube-system/events/kube-dns-v8-1rpl3.13fb8af2d2e1e21c: (12.587148ms) 200 [[kube-scheduler/v1.0.3 (linux/amd64) kubernetes/61c6ac5] 127.0.0.1:42139]
I0818 11:48:31.878725       6 logs.go:41] http: TLS handshake error from 10.245.1.3:47710: remote error: bad certificate
I0818 11:48:32.261033       6 logs.go:41] http: TLS handshake error from 10.245.1.3:47711: remote error: bad certificate
I0818 11:48:32.261612       6 logs.go:41] http: TLS handshake error from 10.245.1.3:47712: remote error: bad certificate
I0818 11:48:32.265000       6 logs.go:41] http: TLS handshake error from 10.245.1.3:47713: remote error: bad certificate
```

Is it likely that the error above is what's preventing the vagrant script from passing the 'Waiting for each minion to be registered with cloud provider' message?

I've had a look at previous issues and found a few similar but they have been closed or related to ansible.  Is there an extra step I need to take or something I'm doing wrong?

**Comment by gauthierbl** [comment]
@rawlingsj I'm also experiencing these same TLS auth errors on the master while using the Vagrant scripts.

I'm also seeing the following on the minion from `/var/log/salt/minion`: 

```
2015-08-18 15:55:13,377 [salt.minion      ][INFO    ][29880] Starting a new job with PID 29880
2015-08-18 15:55:13,378 [salt.utils.lazy  ][DEBUG   ][29880] LazyLoaded grains.items
2015-08-18 15:55:13,380 [salt.minion      ][INFO    ][29880] Returning information for job: 20150818155513360453
2015-08-18 15:55:13,380 [salt.crypt       ][DEBUG   ][29880] Re-using SAuth for ('/etc/salt/pki/minion', 'kubernetes-minion-1', 'tcp://10.245.1.2:4506')
2015-08-18 15:55:13,556 [salt.minion      ][INFO    ][2582] User vagrant Executing command grains.items with jid 20150818155513542857
2015-08-18 15:55:13,556 [salt.minion      ][DEBUG   ][2582] Command details {'tgt_type': 'glob', 'jid': '20150818155513542857', 'tgt': '*', 'ret': '', 'user': 'vagrant', 'arg': [], 'fun': 'grains.items'}
... [loops with a differnt PID] ...
```

The contents of `/etc/salt/pki/minion` are below:

```
[vagrant@kubernetes-minion-1 salt]$ sudo ls -alh /etc/salt/pki/minion
total 20K
drwx------. 2 root root 4.0K Aug 18 14:57 .
drwxr-xr-x. 3 root root 4.0K Aug 18 14:57 ..
-rw-r--r--. 1 root root  451 Aug 18 14:57 minion_master.pub
-r--------. 1 root root 1.7K Aug 18 14:57 minion.pem
-rw-r--r--. 1 root root  451 Aug 18 14:57 minion.pub
```

On the master in `/var/lib/kubelet/kubeconfig1` and on the minion in `/var/lib/kube-proxy/kubeconfig` I have the following which seems relevant. I have changed the value of `insecure-skip-tls-verify` from "true" to "True" with no results.

```
clusters:
- name: local
  cluster:
    insecure-skip-tls-verify: true
```

Not sure if this helps at all.

**Comment by albertsj1** [comment]
I just downloaded the tarball for 1.0.3 and following the [getting started guide for vagrant](http://kubernetes.io/v1.0/docs/getting-started-guides/vagrant.html).  I'm experiencing the same issue as @rawlingsj and see the same messages in the log file.
I'm completely new to Kubernetes and just trying to launch a local cluster with vagrant to get familiar with it.  Is there a release that works properly with Vagrant following the getting started guide?

**Comment by jonmoter** [comment]
+1, seeing the same issue, including the same output in `master:/var/log/kube-apiserver.log` and `minion-1:/var/log/salt/minion`

Kubernetes 1.0.0
Vagrant 1.7.4
VirtualBox 5.0.2

**Comment by invernizzie** [comment]
+1
Kubernetes 1.0.3
Vagrant 1.7.4
VirtualBox 4.3.28

**Comment by pesho** [comment]
+1
Kubernetes 1.0.3
Vagrant 1.7.4
VirtualBox 5.0.0

**Comment by finn0013** [comment]
+1 - this may be the same issue: https://github.com/kubernetes/kubernetes/issues/12892

**Comment by javapapo** [comment]
+1
Kubernetes 1.0.3
Vagrant 1.7.4
VirtualBox 5.0.1 or VirtualBox 4.3.x or

**Comment by finn0013** [comment]
This post will tell how to get Kubernetes/Vagrant running, short term, if you are having this problem:
http://stackoverflow.com/questions/32062868/kubernetes-with-vagrant-nodes-wont-register

**Comment by javapapo** [comment]
Small comment version 1.0.2 works as well, as long as you perform the small change as indicated [here] (https://github.com/kubernetes/kubernetes/pull/12237/files)

**Comment by jfchevrette** [comment]
+1 I'm unable to get 1.0.3/1.0.4 running because nodes will not register

```
[root@kubernetes-master ~]# tail /var/log/kube-apiserver.log
I0826 12:27:06.412600       5 logs.go:41] http: TLS handshake error from 10.245.1.3:39798: remote error: bad certificate
I0826 12:27:07.276523       5 logs.go:41] http: TLS handshake error from 10.245.1.3:39801: remote error: bad certificate
I0826 12:27:07.279504       5 logs.go:41] http: TLS handshake error from 10.245.1.3:39800: remote error: bad certificate
I0826 12:27:07.280921       5 logs.go:41] http: TLS handshake error from 10.245.1.3:39799: remote error: bad certificate
I0826 12:27:07.988666       5 handlers.go:137] GET /api: (39.988µs) 200 [[kubectl/v1.0.3 (darwin/amd64) kubernetes/61c6ac5] 10.245.1.1:51506]
I0826 12:27:08.183967       5 handlers.go:137] GET /api: (229.207µs) 200 [[kubectl/v1.0.3 (darwin/amd64) kubernetes/61c6ac5] 10.245.1.1:51507]
I0826 12:27:08.288998       5 logs.go:41] http: TLS handshake error from 10.245.1.3:39803: remote error: bad certificate
I0826 12:27:08.289397       5 logs.go:41] http: TLS handshake error from 10.245.1.3:39804: remote error: bad certificate
I0826 12:27:08.291977       5 logs.go:41] http: TLS handshake error from 10.245.1.3:39802: remote error: bad certificate
I0826 12:27:08.380730       5 handlers.go:137] GET /api/v1/nodes: (1.867998ms) 200 [[kubectl/v1.0.3 (darwin/amd64) kubernetes/61c6ac5] 10.245.1.1:51508]
```

**Comment by derekwaynecarr** [comment]
I need to backport the following PR into 1.0.x branches:

https://github.com/kubernetes/kubernetes/pull/12514

Apologies to those that lost time to this problem.

The fix is merged at head, and vagrant is stable there if folks want to work from the master branch in the interim.

**Comment by jfchevrette** [comment]
wow I was looking at that yaml file this morning and couldn't see what the issue was. Thanks @derekwaynecarr !

**Comment by rawlingsj** [comment]
I'm not able to test master as it requires boot2docker which in turn causes me issues with incompatible docker client versions.  I'll probably have to wait for the next release but happy to close the issue if someone tests it ok.

**Comment by stuart-warren** [comment]
@rawlingsj see #13095 or #12982 for docker-machine changes

**Comment by derekwaynecarr** [comment]
Closing issue as referenced PR was merged.  This will be available in 1.0.5 release.
